diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0a32ee4e..4200f29c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -387,7 +387,9 @@ install(FILES
     "${CMAKE_CURRENT_BINARY_DIR}/${QUOTIENT_LIB_NAME}/${QUOTIENT_LIB_NAME}ConfigVersion.cmake"
     DESTINATION ${CMakeFilesLocation}
 )
-install(EXPORT_ANDROID_MK ${QUOTIENT_LIB_NAME}Targets DESTINATION ${CMAKE_INSTALL_DATADIR}/ndk-modules)
+if (ANDROID)
+    install(EXPORT_ANDROID_MK ${QUOTIENT_LIB_NAME}Targets DESTINATION ${CMAKE_INSTALL_DATADIR}/ndk-modules)
+endif ()
 
 if (WIN32)
     install(FILES mime/packages/freedesktop.org.xml DESTINATION mime/packages)
