--- DVDStyler-1.8.1/src/mediatrc_ffmpeg.cpp	2011-08-29 19:24:10.000000000 +0200
+++ DVDStyler-1.8.1-new/src/mediatrc_ffmpeg.cpp	2011-08-29 19:24:56.000000000 +0200
@@ -2242,7 +2242,7 @@
 			} else
 				big_picture.quality = (int) ost->st->quality;
 			if (!me_threshold)
-				big_picture.pict_type = 0;
+				big_picture.pict_type = AV_PICTURE_TYPE_NONE;
 //            big_picture.pts = AV_NOPTS_VALUE;
 			big_picture.pts= ost->sync_opts;
 //            big_picture.pts= av_rescale(ost->sync_opts, AV_TIME_BASE*(int64_t)enc->time_base.num, enc->time_base.den);
