--- Makefile	2012-11-19 03:34:12.086348075 +0100
+++ neu.makefile	2012-11-19 04:06:51.615850586 +0100
@@ -68,13 +68,15 @@
 #	scp ${DISTDIR}.tar.gz alex@x37v.info:x37v.info/jack_cpp/code
 
 install: ${LIBNAME}
-	cp ${LIBNAME} ${PREFIX}/lib/
-	cp include/* ${PREFIX}/include/
+	mkdir -p ${DESTDIR}${PREFIX}/lib/
+	mkdir -p ${DESTDIR}${PREFIX}/include/
+	cp ${LIBNAME} ${DESTDIR}${PREFIX}/lib/
+	cp include/* ${DESTDIR}${PREFIX}/include/
 
 uninstall:
-	rm -f ${PREFIX}/lib/${LIBNAME}
+	rm -f ${DESTDIR}${PREFIX}/lib/${LIBNAME}
 	for i in ${INCLUDE_FILES}; do \
-		rm -f ${PREFIX}/$$i; \
+		rm -f ${DESTDIR}${PREFIX}/$$i; \
 	done
 
 clean-swig:
