From 8e3e310a909017373bccca403eefd3f1074edcef Mon Sep 17 00:00:00 2001
From: Gerion Entrup <gerion.entrup@flump.de>
Date: Sat, 22 Mar 2025 10:36:31 +0100
Subject: [PATCH 1/2] make DEV_SANDBOX optional

---
 3party/CMakeLists.txt | 28 +++++++++++++++-------------
 CMakeLists.txt        |  6 +++++-
 2 files changed, 20 insertions(+), 14 deletions(-)

diff --git a/3party/CMakeLists.txt b/3party/CMakeLists.txt
index d5635d56fb..0bae31b6c6 100644
--- a/3party/CMakeLists.txt
+++ b/3party/CMakeLists.txt
@@ -67,18 +67,20 @@ add_subdirectory(vulkan_wrapper)
 if (PLATFORM_DESKTOP)
   add_subdirectory(libtess2)
 
-  set(GLFW_BUILD_DOCS OFF CACHE BOOL "")
-  set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "")
-  set(GLFW_BUILD_TESTS OFF CACHE BOOL "")
-  set(GLFW_INSTALL OFF CACHE BOOL "")
-  set(GLFW_VULKAN_STATIC OFF CACHE BOOL "")
-  set(GLFW_BUILD_WAYLAND OFF CACHE BOOL "")
-  # Disable ARC for glfw and re-enable after it because it's globally set in the root CMakeLists.txt
-  set(CMAKE_OBJC_FLAGS "")
-  add_subdirectory(glfw)
-  set_target_properties(glfw PROPERTIES UNITY_BUILD OFF)
-  set_target_properties(glfw PROPERTIES XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC NO)
-  set(CMAKE_OBJC_FLAGS -fobjc-arc)
+  if (DEV_SANDBOX)
+    set(GLFW_BUILD_DOCS OFF CACHE BOOL "")
+    set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "")
+    set(GLFW_BUILD_TESTS OFF CACHE BOOL "")
+    set(GLFW_INSTALL OFF CACHE BOOL "")
+    set(GLFW_VULKAN_STATIC OFF CACHE BOOL "")
+    set(GLFW_BUILD_WAYLAND OFF CACHE BOOL "")
+    # Disable ARC for glfw and re-enable after it because it's globally set in the root CMakeLists.txt
+    set(CMAKE_OBJC_FLAGS "")
+    add_subdirectory(glfw)
+    set_target_properties(glfw PROPERTIES UNITY_BUILD OFF)
+    set_target_properties(glfw PROPERTIES XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC NO)
+    set(CMAKE_OBJC_FLAGS -fobjc-arc)
 
-  add_subdirectory(imgui)
+    add_subdirectory(imgui)
+  endif()
 endif()
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 40a056580b..4582362ac9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -103,6 +103,8 @@ if(${PLATFORM_MAC})
   endif()
 endif()
 
+option(DEV_SANDBOX "Enable compilation of the developer sandbox when on desktop (not QT based)" ON)
+
 # Sanitizer
 if (PLATFORM_DESKTOP)
   # https://clang.llvm.org/docs/UsersManual.html#controlling-code-generation
@@ -397,7 +399,9 @@ if (PLATFORM_DESKTOP)
     add_subdirectory(qt)
     omim_add_tool_subdirectory(skin_generator)
   endif()
-  add_subdirectory(dev_sandbox)
+  if (DEV_SANDBOX)
+    add_subdirectory(dev_sandbox)
+  endif()
 endif()
 
 omim_add_test_subdirectory(qt_tstfrm)
-- 
2.48.1

